### DISCUSSION TITLE
pom.xml for Salesforce PMD - Jenkins

### DISCUSSION THREAD

**Discussion by Satz-Babu** [pmd/pmd#3594] [General]
Could you please help me with the pom.xml such that the jenkins runs the PMD for Salesforce(Apex/Visualforce pages).
I am stuck here.

**Comment by adangel** [comment]
Since you are mentioning `pom.xml`, I assume, your build-tool is Maven. Then there is this example usage: [Analyzing Javascript](https://maven.apache.org/plugins/maven-pmd-plugin/examples/javascriptReport.html).

This is for javascript, so you'll need to adjust it. Apex/Visualforce is not by default a dependency in the maven-pmd-plugin (only java, javascript, jsp are default dependencies), so you'll need to add these first (similar to [Upgrading PMD at Runtime](https://maven.apache.org/plugins/maven-pmd-plugin/examples/upgrading-PMD-at-runtime.html)). Then you need to tell, where the sources are and how the files are named (maven is mostly a java build tool, so it defaults to java).

```xml
  <properties>
    <pmdVersion>6.40.0</pmdVersion>
  </properties>

  <reporting>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-pmd-plugin</artifactId>
        <version>3.15.0</version>
        <configuration>
          <language>apex</language>
          <rulesets>
            <ruleset>/rulesets/apex/quickstart.xml</ruleset>
            <ruleset>TODO/ruleset/for/visualforce.xml</ruleset>
          </rulesets>
          <includes>
            <include>**/*.cls</include>
            <include>**/*.page</include>
          </includes>
          <compileSourceRoots>
            <compileSourceRoot>${basedir}/classes</compileSourceRoot>
            <compileSourceRoot>${basedir}/pages</compileSourceRoot>
          </compileSourceRoots>
        </configuration>
        <reportSets>
          <reportSet>
            <reports>
              <report>pmd</report>
            </reports>
          </reportSet>
        </reportSets>
        <dependencies>
            <dependency>
              <groupId>net.sourceforge.pmd</groupId>
              <artifactId>pmd-core</artifactId>
              <version>${pmdVersion}</version>
            </dependency>
            <dependency>
              <groupId>net.sourceforge.pmd</groupId>
              <artifactId>pmd-java</artifactId>
              <version>${pmdVersion}</version>
            </dependency>
            <dependency>
              <groupId>net.sourceforge.pmd</groupId>
              <artifactId>pmd-javascript</artifactId>
              <version>${pmdVersion}</version>
            </dependency>
            <dependency>
              <groupId>net.sourceforge.pmd</groupId>
              <artifactId>pmd-jsp</artifactId>
              <version>${pmdVersion}</version>
            </dependency>
            <dependency>
              <groupId>net.sourceforge.pmd</groupId>
              <artifactId>pmd-apex</artifactId>
              <version>${pmdVersion}</version>
            </dependency>
            <dependency>
              <groupId>net.sourceforge.pmd</groupId>
              <artifactId>pmd-visualforce</artifactId>
              <version>${pmdVersion}</version>
            </dependency>
        </dependencies>
      </plugin>
    </plugins>
  </reporting>
```

Notes:
* I didn't try this. It might not work.
* For Apex, I used the quickstart ruleset. This might not be what you want.
* You'll need to provide a ruleset with the rules you want to use for visualforce pages. See [Making Rulesets](https://pmd.github.io/latest/pmd_userdocs_making_rulesets.html) and [Visualforce Rule Index](https://pmd.github.io/latest/pmd_rules_vf.html).
* Maybe you want to fail the build, if there are violations. Then you don't configure a reporting section (which is executed only for the maven site), but you want to do this: [Violation Checking](https://maven.apache.org/plugins/maven-pmd-plugin/examples/violationChecking.html)
* Maybe you create a maven site, then you can a very similar configuration like in "Violation Checking". Instead of the goal "check", use "pmd" and instead of "cpd-check" use "cpd". This will only create the reports in `target/pmd.xml` and `target/cpd.xml`, but not fail the build.
* Maybe you need a Jenkins plugin to pick up these files (target/pmd.xml) - probably this one should work: https://plugins.jenkins.io/warnings-ng/

Since I don't think that maven is the default build tool for apex, maybe it's easier to just use the [PMD CLI](https://pmd.github.io/latest/pmd_userdocs_cli_reference.html). You'd need to figure out, how PMD is installed on your jenkins instance and the add a little script to your build plan to execute PMD. This will generate the file "pmd.xml", which can then be analyzed and visualized by warnings-ng.
